src/aspire/basis/fle_2d.py

```diff
@@ -0,0 +1,746 @@
+        self._calc_k_max()
+
+        if self.match_fb:
+            # Use FB2D and FFB2D heuristic for computing max basis functions
+            self.max_basis_functions = self.k_max[0] + sum(2 * self.k_max[1:])
+        else:
+            # Regular Fourier-Bessel bandlimit (equivalent to pi*R**2)
+            # Final self.count will be < self.max_basis_functions
+            # See self._threshold_basis_functions()
+            self.max_basis_functions = int(self.nres**2 * np.pi / 4)
+
+        self._compute_maxitr_and_numsparse()
+
+        self._compute_cartesian_gridpoints()
+
+        self._precomp()
+
+        # Steerable basis indices
+        self._build_indices()
+
+        # FB compatability indices
+        self._get_fb_compat_indices()
+
+    def _build_indices(self):
+        self.angular_indices = np.abs(self.ells)
+        self.radial_indices = self.ks - 1
+        self.signs_indices = np.array(list(map(fle_ell_sign, self.ells)))
+
+    def indices(self):
+        """
+        Return the precomputed indices for each basis function.
+        """
+        return {
+            "ells": self.angular_indices,
+            "ks": self.radial_indices,
+            "sgns": self.signs_indices,
+        }
+
+    def _get_fb_compat_indices(self):
+        """
+        Generate indices to shuffle basis function ordering and flip signs in order
+            to match `FBBasis2D`.
+        """
+        ind = self.indices()
+        # basis function ordering
+        self.fb_compat_indices = np.lexsort((ind["ks"], ind["sgns"], ind["ells"]))
+        # flip signs
+        self.flip_sign_indices = np.where(ind["sgns"] == 1)
+
+    def _precomp(self):
+        """
+        Precompute the basis functions and other objects used in the evaluation of
+            coefficients.
+        """
+
+        # Find bessel functions zeros (the eigenvalues of the Laplacian on
+        # the disk) and generate the FLE Basis functions
+        self._lap_eig_disk()
+
+        # Some important constants
+        self.smallest_lambda = np.min(self.bessel_zeros)
+        self.greatest_lambda = np.max(self.bessel_zeros)
+        self.max_ell = np.max(np.abs(self.ells))
+        self.h = 1 / (self.nres // 2)
+
+        # give each ell a positive index increasing first in |ell|
+        # then in sign, e.g. 0->1, -1->2, 1->3, -2->4, 2->5, etc.
+        self.ells_p = 2 * np.abs(self.ells) - (self.ells < 0)
+        self.ell_p_max = np.max(self.ells_p)
+        # idx_list[k] contains the indices j of ells_p where ells_p[j] = k
+        idx_list = [[] for i in range(self.ell_p_max + 1)]
+        for i in range(self.count):
+            ellp = self.ells_p[i]
+            idx_list[ellp].append(i)
+        self.idx_list = idx_list
+
+        # real <-> complex
+        self.c2r = precomp_transform_complex_to_real(self.ells)
+        self.r2c = sparse.csr_matrix(self.c2r.transpose().conj())
+
+        # create self.nus to be able to access the physical value of ell,
```

Comment:
It's used in `step2_t` and `step2`. It's basically a remapping of the ells to positive integers. I'll add a better comment

Reply:
nus is an ordered list of just the ells with no repeats. it's used farther down in `step2_t` and `step2`. I'll add a better comment

Reply:
interesting. Does it need to get reordered with any of the index gymnastics we do? (I'm thinking no, because you mention reordering at certain times, but want to confirm).